File: Java/Longest-Substring-Without-Repeating-Characters.java
public class Solution {
    /**
     * @param s: a string
     * @return: an integer
     */
    public int lengthOfLongestSubstring(String s) {
        // write your code here
        int maxlen = 0, left = 0;
        int []hash = new int[256];
        for (int i = 0; i < s.length(); i ++) {
            hash[s.charAt(i)] ++;
            while (hash[s.charAt(i)] > 1) {
                hash[s.charAt(left)] --;
                left ++;
            }
            maxlen = Math.max(maxlen, i - left + 1);
        }
        return maxlen;
    }
}

File: Cpp/Longest-Substring-Without-Repeating-Characters.cpp
#include <string>

class Solution {
public:
    /**
     * @param s: a string
     * @return: an integer 
     */
    int lengthOfLongestSubstring(string s) {
        // write your code here
        const int ASCII_MAX = 256;
        int appeared_count[256];
        fill(appeared_count, appeared_count + ASCII_MAX, 0);
        int maxlen = 0, start = 0;
        for (int i = 0; i < s.size(); i ++) {
            char ch = s[i];
            if (appeared_count[ch] == 0) {
                appeared_count[ch] ++;
                maxlen = max(maxlen, i - start + 1);
            } else {
                while (appeared_count[ch] > 0) {
                    appeared_count[s[start]] --;
                    start ++;
                }
                appeared_count[ch] ++;
            }
        }
        return maxlen;
    }
};
